rust/src/operations/filesystem_check.rs

```diff
@@ -0,0 +1,221 @@
+//! Audit the Delta Table for active files that do not exist in the underlying filesystem and remove them.
+//!
+//! Active files are ones that have an add action in the log, but no corresponding remove action.
+//! This operation creates a new transaction containing a remove action for each of the missing files.
+//!
+//! This can be used to repair tables where a data file has been deleted accidentally or
+//! purposefully, if the file was corrupted.
+//!
+//! # Example
+//! ```rust ignore
+//! let mut table = open_table("../path/to/table")?;
+//! let (table, metrics) = FileSystemCheckBuilder::new(table.object_store(), table.state).await?;
+//! ````
+use crate::action::{Action, Add, DeltaOperation, Remove};
+use crate::operations::transaction::commit;
+use crate::storage::DeltaObjectStore;
+use crate::table_state::DeltaTableState;
+use crate::DeltaDataTypeVersion;
+use crate::{DeltaDataTypeLong, DeltaResult, DeltaTable, DeltaTableError};
+use futures::future::BoxFuture;
+use futures::StreamExt;
+pub use object_store::path::Path;
+use object_store::ObjectStore;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::SystemTime;
+use std::time::UNIX_EPOCH;
+use url::{ParseError, Url};
+
+/// Audit the Delta Table's active files with the underlying file system.
+/// See this module's documentaiton for more information
+#[derive(Debug)]
+pub struct FileSystemCheckBuilder {
+    /// A snapshot of the to-be-checked table's state
+    state: DeltaTableState,
+    /// Delta object store for handling data files
+    store: Arc<DeltaObjectStore>,
+    /// Don't remove actions to the table log. Just determine which files can be removed
+    dry_run: bool,
+}
+
+/// Details of the FSCK operation including which files were removed from the log
+#[derive(Debug)]
+pub struct FileSystemCheckMetrics {
+    /// Was this a dry run
+    pub dry_run: bool,
+    /// Files that wrere removed successfully
+    pub files_removed: Vec<String>,
+}
+
+struct FileSystemCheckPlan {
+    /// Version of the snapshot provided
+    version: DeltaDataTypeVersion,
+    /// Delta object store for handling data files
+    store: Arc<DeltaObjectStore>,
+    /// Files that no longer exists in undlying ObjectStore but have active add actions
+    pub files_to_remove: Vec<Add>,
+}
+
+fn is_absolute_path(path: &str) -> DeltaResult<bool> {
+    match Url::parse(path) {
+        Ok(_) => Ok(true),
+        Err(ParseError::RelativeUrlWithoutBase) => Ok(false),
+        Err(_) => Err(DeltaTableError::Generic(format!(
+            "Unable to parse path: {}",
+            &path
+        ))),
+    }
+}
```

Comment:
I think this will not work for local paths. One pattern we apply in `object_store` and here is to try and cannonicalize the path and if it fails proceed with url parsing. e.g.

https://github.com/delta-io/delta-rs/blob/a60129bb48a24f94e8312c12940f6ec96685af05/rust/src/builder.rs#L355-L358

Reply:
I don't follow on why it would not work on local paths. Paths in the Delta log must follow [rfc2396](https://www.ietf.org/rfc/rfc2396.txt) which states absolute paths must have a scheme.
Do you have a particular example that would cause the unit test for this function to fail?

Reply:
well  .. in that case I am wrong! If that paths are required to have a scheme, then url parsing should always work for valid paths ... The windows case is also covered in your tests, so all is well :).